### PR TITLE
Don't hoist ref locals

### LINKED ISSUE #40251: error MSB6006: "csc.exe" exited with code -2146232797.
Came across this crashing issue while working on Advent of Code 2019. I am using a few C# 8 features in this file, including switch expressions and ref returns. Stacktrace from https://sharplab.io is pasted below.

**Version Used**: 
```
C:\Microsoft Visual Studio\2019\dogfood\Enterprise>csc
Microsoft (R) Visual C# Compiler version 3.5.0-beta2-19606-01 (e8ab1541)
Copyright (C) Microsoft Corporation. All rights reserved.
```
**Steps to Reproduce**:

1. Compile the following code
```cs
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AdventOfCode._2019
{
    class IntCode
    {
        enum Op : long
        {
            Add = 1,
            Mul = 2,
            In = 3,
            Out = 4,
            JumpTrue = 5,
            JumpFalse = 6,
            LessThan = 7,
            Equals = 8,
            Base = 9,
            Halt = 99,
        }

        enum Mode : long
        {
            Pos = 0,
            Imm = 1,
            Relative = 2
        }

        public delegate Task<long> ReadInput();
        public delegate void WriteOutput(long value);

        public long PC { get; set; } = 0;

        public long RelativeBase { get; set; } = 0;

        public bool IsHalt { get; private set; } = false;

        public long this[int index]
        {
            get => ReadMemory(index);
            set => ReadMemory(index) = value;
        }

        public ReadInput Reader { get; set; }
        public WriteOutput Writer { get; set; }

        public IntCode(IEnumerable<int> memory)
            : this(memory, null, null)
        { 
        }

        public IntCode(IEnumerable<int> program, ReadInput reader, WriteOutput writer)
            : this(program.Select(i => (long)i), reader, writer)
        {
        }

        public IntCode(IEnumerable<long> memory, ReadInput reader, WriteOutput writer)
        {
            _memory = memory.ToArray();
            Reader = reader;
            Writer = writer;
        }

        public async Task Step()
        {
            if (!IsHalt)
            {
                (Op op, Mode[] modes) = Decode();

                if (op == Op.Halt)
                    IsHalt = true;
                else if (op == Op.In)
                    ReadMemory(ReadPC()) = await Reader();
                else if (op == Op.Out)
                    Writer(ReadArg(modes[0]));
                else if (op == Op.Base)
                    RelativeBase += ReadArg(modes[0]);
                else
                {
                    long val1 = ReadArg(modes[0]);
                    long val2 = ReadArg(modes[1]);

                    if (op == Op.JumpTrue || op == Op.JumpFalse)
                    {
                        if ((op == Op.JumpTrue && val1 != 0) || (op == Op.JumpFalse && val1 == 0))
                            PC = val2;
                    }
                    else
                    {
                        ReadMemory(ReadPC()) = op switch
                        {
                            Op.Add => val1 + val2,
                            Op.Mul => val1 * val2,
                            Op.LessThan => val1 < val2 ? 1 : 0,
                            Op.Equals => val1 == val2 ? 1 : 0,
                            _ => throw new InvalidOperationException()
                        };
                    }
                }
            }
        }

        public async Task Run()
        {
            while (!IsHalt)
            {
                await Step();
            }
        }

        (Op, Mode[]) Decode()
        {
            long instr = ReadPC();
            
            Op op = (Op)(instr % 100);
            instr /= 100;

            Mode[] modes = new Mode[3];
            modes[0] = (Mode)(instr % 10);
            instr /= 10;
            modes[1] = (Mode)(instr % 10);
            instr /= 10;
            modes[2] = (Mode)(instr % 10);

            return (op, modes);
        }

        private long ReadArg(Mode mode) => mode switch
        {
            Mode.Imm => ReadPC(),
            Mode.Pos => ReadMemory(ReadPC()),
            Mode.Relative => ReadMemory(ReadPC() + RelativeBase),
            _ => throw new InvalidOperationException()
        };

        private long ReadPC()
        {
            return ReadMemory(PC++);
        }

        private ref long ReadMemory(long address)
        {
            if (address >= _memory.Length)
            {
                long[] newMemory = new long[_memory.Length * 2];
                for (int i = 0; i < _memory.Length; i++)
                    newMemory[i] = _memory[i];

                _memory = newMemory;
            }
            return ref _memory[address];
        }

        private long[] _memory;
    }
}
```

**Expected Behavior**:
csc.exe does not crash

**Actual Behavior**:
csc.exe crashes

https://sharplab.io/ show's an interesting stacktrace for this issue:
```
System.InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Syntax.SwitchExpressionSyntax' to type 'Microsoft.CodeAnalysis.CSharp.Syntax.AwaitExpressionSyntax'.
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.HoistRefInitialization(SynthesizedLocal local, BoundAssignmentOperator node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.VisitAssignmentOperator(BoundAssignmentOperator node)
   at Microsoft.CodeAnalysis.CSharp.BoundAssignmentOperator.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.AsyncMethodToStateMachineRewriter.VisitExpressionStatement(BoundExpressionStatement node)
...
```

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #40251

@gafter could you please take a look? Without this change, we crash in debug mode, while giving the "expected" behavior (either giving an emit diagnostic or compiling successfully) in release mode.

### CODE REVIEW

**Comment by jcouv** [comment]
Glad it worked :-)
Can you take a look at linked issues? I think there was more than one related to ref locals in async methods.

**Comment by RikkiGibson** [comment]
I poked around at some linked issues and didn't see others that seemed affected. I think this only resolves issues where either:

1. the compiler was crashing in lowering,
2. or there were errors given that should not have been.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs [hunk 1]
```diff
@@ -5606,5 +5606,60 @@ static async Task Main()
...
+            _ => throw null
+        };
+    }
+
+    private ref long ReadMemory() => throw null;
+}
+";
+
+            var comp = CreateCompilation(source, options: TestOptions.DebugDll);
+            comp.VerifyEmitDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs:5632]: >comp.VerifyEmitDiagnostics(); [](start = 12, length = 29)

Consider running the scenario and observing expected behavior at runtime.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs [hunk 2]
```diff
@@ -5606,5 +5606,60 @@ static async Task Main()
...
+            comp = CreateCompilation(source, options: TestOptions.ReleaseDll);
+            comp.VerifyEmitDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs:5634]: >comp.VerifyEmitDiagnostics(); [](start = 12, length = 29)

Here as well.

**Comment by jcouv** [comment]
@RikkiGibson Here are some issues I had in mind:
- https://github.com/dotnet/roslyn/issues/42738
- https://github.com/dotnet/roslyn/issues/40223

They also involve ref locals in async methods, but those aren't crashes, so unlikely the same issue :-(

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs [hunk 3]
```diff
@@ -5606,5 +5606,87 @@ static async Task Main()
...
+        public void AssignRefWithAwait()
+        {
+            const string source = @"
+using System.Threading.Tasks;
+
+class IntCode
+{
+    public async Task Step(Task<int> t)
+    {
+        ReadMemory() = await t;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs:5674]: Consider adding a test where we have a `ref` l-value and use a compound operator. Essentially 

```cs
ReadMemory() += await t;
```

**Pull request merged**